### PR TITLE
Begin abstraction of wallet syncing out of the wallet package.

### PR DESCRIPTION
**Pull request by jrick**
This change begins the removal of the dcrd JSON-RPC client dependency
from the wallet package and type.  All notification processing to keep
the wallet up to date has been moved to the chain package.

This change is being performed to make switching to other network
syncing backends easier in the future.  Most importantly, it is
required for any kind of support for both RPC and SPV network
backends.

To avoid some churn, the wallet must still be associated with an
object that provides basic network operations such as loading data
filters, fetching headers, and transaction publishing.  This object is
now an interface type rather than the concrete RPC client type.  At
the moment, this interface is currently too narrow to allow an SPV
backend as a drop in replacement, but the goal is to remove methods
from it until this becomes possible.

There are still some features in the wallet package which do require
the RPC client.  These features either take the RPC client as a method
parameter or assert that the network backend is the RPC client.

This change also uses this API break as an opportunity to begin
passing around context.Context for request-scoped cancellation and
timeouts.  The methods of the network backend interface all take a
context, and while it is not used by the backend, they will be used
for any future implementations.

### CODE REVIEW

**Comment by jrick** [comment]
This will need to wait until decred/dcrrpcclient#94 is merged so that shutdown while attempting the dcrd RPC connection doesn't hang.  On master, this behavior was non encountered since the client connection handling was performed asynchronous to the main function.

**Comment by jrick** [comment]
Rebased over master to fix merge conflicts after the repo layout changes and ready for review.

File: dcrwallet.go [hunk 1]
```diff
@@ -142,86 +196,78 @@ func walletMain() error {
...
-				var unlockAfter <-chan time.Time
-				err = w.Unlock(passphrase, unlockAfter)
-				if err != nil {
-					log.Errorf("Incorrect passphrase in pass config setting.")
-					return err
-				}
-			} else {
-				passphrase = startPromptPass(w)
+		if cfg.Pass != "" {
+			err = w.Unlock([]byte(cfg.Pass), nil)
```
Comment (jolan) [inline dcrwallet.go]: You need to add back:

```go
passphrase = []byte(cfg.Pass)
```

otherwise ticket buying with cfg.Pass will be broken.

**Comment by jrick** [comment]
@alexlyp i know you already checked this with an earlier version but I agree that it would be good to get it tested with decrediton again.

**Pull request merged**